pkg/controller/master/addon/common.go

```diff
@@ -0,0 +1,402 @@
+package addon
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"time"
+
+	helmv1 "github.com/k3s-io/helm-controller/pkg/apis/helm.cattle.io/v1"
+	"github.com/rancher/wrangler/pkg/condition"
+
+	"github.com/sirupsen/logrus"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	"github.com/harvester/harvester/pkg/util"
+)
+
+const (
+	// addon enqueue self interval, defaults to 5s
+	enqueueInterval = 5
+	// addon operation timeout, defaults to 15 min as cluster fisrt setup takes long time, range: [1 (min) ~ 60 (min)]
+	addonTimeout    = 15
+	addonTimeoutMin = 1
+	addonTimeoutMax = 60
+)
+
+type addonOperationRecord struct {
+	Operation                harvesterv1.AddonOperation `json:"operation,omitempty"`
+	OperationTimestamp       metav1.Time                `json:"operationTimestamp,omitempty"`
+	OperationTimestampString string                     `json:"operationTimestampstr,omitempty"`
+}
+
+func newAddonOperationRecord(operation string, tm string) (*addonOperationRecord, error) {
+	aor := addonOperationRecord{}
+
+	if operation == string(harvesterv1.AddonEnableOperation) {
+		aor.Operation = harvesterv1.AddonEnableOperation
+	} else if operation == string(harvesterv1.AddonDisableOperation) {
+		aor.Operation = harvesterv1.AddonDisableOperation
+	} else if operation == string(harvesterv1.AddonUpdateOperation) {
+		aor.Operation = harvesterv1.AddonUpdateOperation
+	} else {
+		return nil, fmt.Errorf("invalid operation %s new addonOperationRecord", operation)
+	}
+
+	t, err := time.Parse(time.RFC3339, tm)
+	if err != nil {
+		return nil, fmt.Errorf("invalid time %s to new addonOperationRecord", tm)
+	}
+
+	aor.OperationTimestamp.Time = t
+	aor.OperationTimestampString = tm
+
+	return &aor, nil
+}
+
+func (h *Handler) updateAddonConditions(aObj *harvesterv1.Addon, cond condition.Cond, status corev1.ConditionStatus, reason, message string, updateLastTransitionTime bool) {
+	// wrangler's cond function is not very fit for here, any read/write needs to loop; play condition directly
+	for i := range aObj.Status.Conditions {
```

Comment:
it can set one field at a time, and it will always loop to find this find; when you update 5 fields in a function, the wrangler function will blindly loop 5 times ...

Reply:
and, IIRC, wrangler does not have function to update the `LastTransitTime` (or `LastUpdateTime`), it only has one public function to update one of those 2 timestamps.